### PR TITLE
 Add a `fips` flag to install-config.yaml

### PR DESCRIPTION
**Pull request by cgwalters**
Part of: openshift/enhancements#15

We added FIPS to the MCO a while ago:
openshift/machine-config-operator#889

However, during some discussion it became clear that the main
use case for FIPS is "day 1" - it doesn't make sense to turn it
on "day 2" because the standard requires that e.g. long-term key
material was created with FIPS enabled.

Further, it's unlikely that admins will want to turn it *off*
if they ever had it on.

This is a good candidate for an install config.

### CODE REVIEW

File: pkg/types/installconfig.go [hunk 1]
```diff
@@ -105,6 +105,9 @@ type InstallConfig struct {
 	// When no strategy is specified, the strategy is `External`.
 	// +optional
 	Publish PublishingStrategy `json:"publish,omitempty"`
+
+	// FIPS configures https://en.wikipedia.org/wiki/Federal_Information_Processing_Standards
```
Comment (wking) [inline pkg/types/installconfig.go]: nit: https://www.nist.gov/itl/fips-general-information seems like a more canonical URI?

File: pkg/types/installconfig.go [hunk 2]
```diff
@@ -105,6 +105,9 @@ type InstallConfig struct {
...
+	FIPS *bool `json:"fips,omitempty"`
```
Comment (abhinavdahiya) [inline pkg/types/installconfig.go]: Why boolean ptr ?

Also why is this not a installer-machine-pool configuration like hyperthreading?

File: pkg/asset/machines/machineconfig/manifest.go [hunk 3]
```diff
@@ -22,27 +22,26 @@ var (
...
+	var ret []*asset.File
+	for i, c := range configs {
 		if c == nil {
 			continue
 		}
 		configData, err := yaml.Marshal(c)
 		if err != nil {
 			return nil, err
 		}
-		data = append(data, []byte("---\n")...)
```
[the remark this answers is not in the document]
Comment (cgwalters) [reply pkg/asset/machines/machineconfig/manifest.go]: Hum...is this for something like Terraform internal dependency tracking?

File: pkg/types/installconfig.go (hunk 2 above)
Comment (cgwalters) [reply pkg/types/installconfig.go]: I couldn't think of a use case for a "partially FIPS" cluster, whereas it's a clear use case to e.g. just disable hyperthreading on workers, or a specific pool of workers.

(That said I think what we really want is something more like https://lwn.net/Articles/799454/ where "hyperthreading" is determined automatically - we have all of the info for it.  But anyways)

**Comment by cgwalters** [comment]
To emphasize, this PR is just "sugar" for providing MachineConfig objects which set `fips: true` via [additional manifests](https://github.com/openshift/installer/blob/master/docs/user/customization.md#install-time-customization-for-machine-configuration).

**Comment by cgwalters** [comment]
OK, fixed the unit tests, and addressed comments!

File: pkg/asset/machines/machineconfig/fips.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+// ForFIPSEnabled creates the MachineConfig to enable FIPS.
+// See also https://github.com/openshift/machine-config-operator/pull/889
+func ForFIPSEnabled(role string) *mcfgv1.MachineConfig {
+	return &mcfgv1.MachineConfig{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "machineconfiguration.openshift.io/v1",
+			Kind:       "MachineConfig",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("01-%s-fips", role),
```
Comment (abhinavdahiya) [inline pkg/asset/machines/machineconfig/fips.go]: The rest of MCs are 99-

File: pkg/asset/machines/master.go [hunk 5]
```diff
@@ -315,6 +315,10 @@ func (m *Master) Generate(dependencies asset.Parents) error {
 	if ic.SSHKey != "" {
 		machineConfigs = append(machineConfigs, machineconfig.ForAuthorizedKeys(ic.SSHKey, "master"))
 	}
+	if ic.FIPS {
+		machineConfigs = append(machineConfigs, machineconfig.ForFIPSEnabled("master"))
+		fmt.Println("Enabling master fips")
```
Comment (abhinavdahiya) [inline pkg/asset/machines/master.go]: Drop the printing

**Comment by abhinavdahiya** [comment]
/hold

Need to make sure a global FIPS option is more appropriate than per pool.

**Comment by cgwalters** [comment]
Addressed comments!

**Comment by cgwalters** [comment]
> Need to make sure a global FIPS option is more appropriate than per pool.

Fair question - who makes that call?

**Comment by sdodson** [comment]
> Fair question - who makes that call?

Discussed during group G, feedback from PM and @crawford is that non FIPs hosts taint a cluster so it has to be cluster scoped.

**Pull request merged**